### PR TITLE
Move Nimbus.kt to Application Services

### PR DESCRIPTION
**Pull request by jhugman**
This PR is the first of a pair: one in Android Components, the other in Application Services.

It moves the file `Nimbus.kt` and associated test `NimbusTest.kt` from the `mozilla.components.service.nimbus` module to `org.mozilla.experiments.nimbus`.

It should not be landed until the corresponding app-services PR lands. This PR will fix the breaking changes that one will introduce (i.e. duplicate class definition).

For a number of reasons, including code migration, the package, name and methods are maintained across the changes.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by grigoryk** [comment]
Just left a comment in https://github.com/mozilla/application-services/pull/4036.

**Comment by jhugman** [comment]
While this PR isn't ready to land, it may be a good idea to get a big picture view of the intent.

[This document][1] is a good place to discuss how we go from here.

[1]: https://docs.google.com/document/d/1nibSjM9Vslqj6GKVxePq6fkOp4Yg4euzGeVidHBUnC4/edit?usp=sharing

**Comment by jhugman** [comment]
Ready to review, but will not build until https://github.com/mozilla/application-services/pull/4036 is landed, a Application Services release is made, and A-C upgrades to that release.

**Review by travis79 — APPROVED** [review]
Nice!

File: components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt [hunk 1]
```diff
@@ -1,554 +1,85 @@
...
+    observer = Observer(observable),
+    delegate = NimbusDelegate(
+        dbScope = CoroutineScope(Executors.newSingleThreadExecutor(
+            NamedThreadFactory("NimbusDbScope")
+        ).asCoroutineDispatcher()),
+        fetchScope = CoroutineScope(Executors.newSingleThreadExecutor(
+            NamedThreadFactory("NimbusFetchScope")
+        ).asCoroutineDispatcher()),
+        errorReporter = errorReporter,
+        logger = { logger.info(it) }
```
Comment (csadilek) [inline components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt:67]: nit: You used to log errors here in various functions, now it's all infos. Intentional?

**Review by csadilek — APPROVED** [review]
One nit/q below, but this looks OK to me!

File: components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt (hunk 1 above)
Comment (jhugman) [reply components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt:67]: Everything that used to report an error via the `logger.error()` is now doing it via an `errorReporter`. 

The only exception to this is when an exception is caught coming from `errorReporter`. If `errorReporter` is itself erroring, I'm guessing nimbus won't be the only one finding out about it.

If/when `Nimbus` does log non-error messages I'd rather do it by `info` than by `error`.

**Comment by grigoryk** [comment]
Corresponding A-S release is now available, so let's land this.

I've fixed a linter issue, updated appservices version, rebased. Just running a local smoke test in Fenix against this, but good to go otherwise once the PR is green.

**Comment by grigoryk** [comment]
Local Fenix build appears to run well!

However, first hurdle: this PR broke `samples-glean`. Going to try fixing it now.

**Comment by grigoryk** [comment]
Seeing a startup crash in `samples-glean`:
```
05-12 16:08:36.601  8653  8653 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{org.mozilla.samples.glean.debug/org.mozilla.samples.glean.MainActivity}: org.mozilla.experiments.nimbus.internal.NimbusErrorException$DatabaseNotReady: Initialization of the database is not yet complete
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3270)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3409)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:214)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7356)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: Caused by: org.mozilla.experiments.nimbus.internal.NimbusErrorException$DatabaseNotReady: Initialization of the database is not yet complete
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at org.mozilla.experiments.nimbus.internal.NimbusError.intoException(nimbus.kt:956)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at org.mozilla.experiments.nimbus.internal.NimbusClient.getExperimentBranch(nimbus.kt:1557)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at org.mozilla.experiments.nimbus.Nimbus.getExperimentBranch(Nimbus.kt:317)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at org.mozilla.samples.glean.MainActivity.configureButton(MainActivity.kt:103)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at org.mozilla.samples.glean.MainActivity.setupNimbusExperiments(MainActivity.kt:78)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at org.mozilla.samples.glean.MainActivity.onCreate(MainActivity.kt:58)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:7802)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:7791)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1299)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3245)
05-12 16:08:36.601  8653  8653 E AndroidRuntime: 	... 11 more
```

**Comment by grigoryk** [comment]
Well, seems like the stacktrace above is an initialization issue in the sample app. I don't see Fenix crashing locally at all, and did manage to get the list of all known experiments displayed in the hidden nimbus UI within Fenix.

**Comment by grigoryk** [comment]
Well, the toolchain task keeps failing over with this:
```
[task 2021-05-13T00:25:06.917Z] > Task :samples-glean:mergeExtDexRelease
[task 2021-05-13T00:25:07.117Z] 
[task 2021-05-13T00:25:07.118Z] FAILURE: Build failed with an exception.
[task 2021-05-13T00:25:07.118Z] 
[task 2021-05-13T00:25:07.118Z] * What went wrong:
[task 2021-05-13T00:25:07.118Z] Execution failed for task ':samples-glean:mergeReleaseJavaResource'.
[task 2021-05-13T00:25:07.118Z] > A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
[task 2021-05-13T00:25:07.118Z]    > More than one file was found with OS independent path 'META-INF/concept-base_release.kotlin_module'.
```

`:samples-glean:mergeExtDexDebug` does work, though.

**Comment by grigoryk** [comment]
Looking at the dependency graph for the failing `:samples-glean` app, we see this:
```
+--- project :service-nimbus
|    +--- org.mozilla.appservices:nimbus:76.0.0
|    |    +--- org.mozilla.appservices:full-megazord:76.0.0
|    |    |    +--- org.jetbrains.kotlin:kotlin-android-extensions-runtime:1.3.71 -> 1.4.30 (*)
|    |    |    \--- org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.3.71 -> 1.4.30 (*)
|    |    +--- org.jetbrains.kotlin:kotlin-android-extensions-runtime:1.3.71 -> 1.4.30 (*)
|    |    +--- net.java.dev.jna:jna:5.6.0
|    |    +--- androidx.annotation:annotation:1.2.0
|    |    +--- org.jetbrains.kotlinx:kotlinx-coroutines-android:1.4.2 (*)
|    |    +--- org.mozilla.components:service-glean:75.0.0
|    |    |    +--- org.mozilla.telemetry:glean:35.0.0 (*)
|    |    |    +--- org.mozilla.components:concept-fetch:75.0.0
|    |    |    |    \--- org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.4.30 (*)
|    |    |    +--- org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.4.30 (*)
|    |    |    +--- org.jetbrains.kotlinx:kotlinx-coroutines-android:1.4.2 (*)
|    |    |    +--- androidx.work:work-runtime-ktx:2.4.0 (*)
|    |    |    +--- org.mozilla.components:support-ktx:75.0.0
|    |    |    |    +--- org.mozilla.components:support-base:75.0.0
|    |    |    |    |    +--- org.mozilla.components:concept-base:75.0.0
```

Note that `org.mozilla.components:concept-base:75.0.0` is being brought in. This also happens to be quite a bit behind the current version of this repo, 75.0.12. (FWIW, a-c `master` is currently on 90.0.0, but unreleased, only on nightlies).

**Pull request merged**